---
title: "Investigating the algal symbiont's response to a freshwater event"
author: "Simmi Nishad, Isabel Novick, Corinne Vietorisz"
date: "3/29/2021"
output: html_document
---

set your working directory
```{r}
setwd("~/Desktop/BI586/Assign2_proj4")
```

```{r include=FALSE}
knitr::opts_chunk$set(comment = NA)
```

**Version Control:**<br> 
R 1.4.1103<br> 
DESeq2 1.30.1<br>
affycoretools 1.62.0<br>
arrayQualityMetrics 3.46.0<br>
genefilter 1.72.1<br>
Biobase 2.50.0<br>
ggplot2 3.3.3<br>
dplyr 1.0.5<br>
pheatmap 1.0.12<br>
vegan 2.5-7<br>
ggrepel 0.9.1<br>
tidyverse 1.3.0<br> 

**Introduction:** The Flower and Garden Banks has been enduring minor events for decades until July 2016 when a mortality event occurred. After Hurricane Harvey, tissue samples were collected to study the impacts of the freshwater runoff on coral and symbionts. These samples were collected at two time points: September 2017 (stress) and October 2017 (recover) to study the gene expression of Orbicella faveolata and Orbicella franksi. Also studied was Breviolum minutum, an algal symbiont which showed differentially expressed genes (DEGs) at different sampling time points. This may be the algal symbiont's response to the freshwater event (change in habitat). The purpose of our assignment was to study these DEGs at two time points, the stress and recovery. By utilizing arrayQualityMetrics, we were able to identify outliers. DESeq2 was leveraged to identify these DEGs and create a statistical model to study the expression patterns. 



**Conduct array quality metrics to detect and remove outliers**

```{r Packages, message=FALSE, warning=FALSE}
library(DESeq2) 
library(affycoretools)
library(arrayQualityMetrics)
library(genefilter)
library(Biobase)
library(ggplot2)
library(dplyr)
library(pheatmap)
library(vegan)
library(ggrepel)
library(tidyverse)
```

**read in iso2gene tab separated file**
```{r message=FALSE, warning=FALSE}
gg=read.table("Bmin_iso2gene.tab",sep="\t")
```

**read in counts** 
```{r message=FALSE, warning=FALSE}
countData <- read.table("allcounts_Harvey_Sym.txt") 
head(countData)
length(countData[,1])
```


**look for outliers**
```{r message=FALSE, warning=FALSE}
treat=c( "Recovery", "Recovery", "Recovery", "Stress", "Stress", "Stress") #creating a condition table
g=data.frame(treat)
g
colData= g
```

**Convert the counts data, treatment information, and desired statistical model into a DESeq object.**
```{r message=FALSE, warning=FALSE}
dds=DESeqDataSetFromMatrix(countData=countData,
                           colData = g,
                           design = ~treat)

vsd.ge=assay(vst(dds))
rl=vst(dds)
e=ExpressionSet(assay(rl), AnnotatedDataFrame(as.data.frame(colData(rl))))
arrayQualityMetrics(e, intgroup=c("treat"),force=T) 
```


**fav_stressC detected as outliers via the Distances between arrays method (Figure 2) and MA plot method (Figure 9). For the purpose of this assignment, we will include it, but in a real scenario we would exclude it.**


**one step DESeq that does the following: estimating size factors, estimating dispersions, gene-wise dispersion estimates, mean-dispersion relationship, final dispersion estimates, fitting model and testing.**
```{r message=FALSE, warning=FALSE}
dds<-DESeq(dds)
```

**These results allow us to make sure nothing abnormal is going on. The post-normalized counts are pvalues that correspond to see how differentially expressed a gene is between treatments. padj adjusts pvalues for multiple tests. The baseMean is average count for one comparison (whichever treatment group came first). log2FoldChange shows differentiation in counts between the basemean vs other treatment group. The recovery treatment group had avg of 5.57 counts; stress group had a 2fold change (logtransformed).**

```{r message=FALSE, warning=FALSE}
head(dds)
res<- results(dds) 
res
```



**This is the normalization method.**
```{r message=FALSE, warning=FALSE, fig.cap= "**Figure 1.** The dispersion plot visualizes what DEseq2 is doing to our data; fitting it to this curve."}
plotDispEsts(dds, main="Wright et al. 2019 symbiont dispersion plot")
```


**Stress vs recovery pairwise comparisons**

**We can only do pairwise comparisons. Levels are used because gene expression is always relative; the first term is the treatment and the second term is the one that's relative (treatment 1st, control 2nd)**
```{r message=FALSE, warning=FALSE}
colData$Stress<-factor(colData$treat, levels=c("Stress","Recovery"))
```

```{r message=FALSE, warning=FALSE}
resStress <- results(dds, contrast=c("treat","Stress","Recovery"))
```

**False discovery rate pvalues are adjusted to < 10%**
```{r message=FALSE, warning=FALSE}
table(resStress$padj<0.01)
```

**Of the 5093 reads, p-adjusted reads are 6. p-adjusted values are used to look at numbers of differentially expressed genes - here we have 6 diferentially expressed genes.p-adjusted accounts for all the different times things are compared. Without adjustment we could get a false positive by chance.**

```{r message=FALSE, warning=FALSE}
summary(resStress)
```

**A lot of genes were removed due to low counts; 0.1% of genes upregulated in stress relative to recovery, and 0.43% were downregulated under stress relative to recovery**

**This is another way to look at the number of differentially expressed genes (6). The number of significantly differentially expressed genes excluding the no/low count genes.**

```{r message=FALSE, warning=FALSE}
nrow(resStress[resStress$padj<0.1 & !is.na(resStress$padj),])
```
  

```{r message=FALSE, warning=FALSE, fig.cap= "**Figure 2.** MA plot shows the mean of normalized counts. The left is more lowly expressed genes abd right is higher expressed genes. The blue is all the differentially expressed genes. Negative fold change is downregulated genes and postitive fold change is upregulated genes."}
plotMA(resStress, main="Stress vs Recovery", ylim=c(-2,2))
```




**Our results are put in a data frame. 23 are upregulated and 98 are downregulated.**
```{r message=FALSE, warning=FALSE}
results <- as.data.frame(resStress)
head(results)

nrow(resStress[resStress$padj<0.1 & resStress$log2FoldChange > 0 & !is.na(resStress$padj),]) #no. genes upregulated
nrow(resStress[resStress$padj<0.1 & resStress$log2FoldChange < 0 & !is.na(resStress$padj),]) #no. genes downregulated
```


**A supplementary table of info.**
```{r message=FALSE, warning=FALSE}
write.table(resStress, file="StressvRecovery.txt", quote=F, sep="\t")
```

```{r message=FALSE, warning=FALSE}
cd <- read.table("StressvRecovery.txt")
head(cd)

```


