---
title: "Investigating the algal symbiont's response to a freshwater event"
author: "Simmi Nishad, Isabel Novick, Corinne Vietorisz"
date: "3/29/2021"
output: html_document
---

set your working directory
```{r}
setwd("~/Desktop/BI586/Assign2_proj4")
```

```{r include=FALSE}
knitr::opts_chunk$set(comment = NA)
```

**Version Control:**<br> 
R 1.4.1103<br> 
DESeq2 1.30.1<br>
affycoretools 1.62.0<br>
arrayQualityMetrics 3.46.0<br>
genefilter 1.72.1<br>
Biobase 2.50.0<br>
ggplot2 3.3.3<br>
dplyr 1.0.5<br>
pheatmap 1.0.12<br>
vegan 2.5-7<br>
ggrepel 0.9.1<br>
tidyverse 1.3.0<br> 

**Introduction:** The following analyses is based on "Gene Expression of Endangered Coral (Orbicella spp.) in Flower Garden Banks National Marine Sanctuary After Hurricane Harvey" (Wright et. al. 2019). After Hurricane Harvey, tissue samples were collected to study the impacts of the freshwater runoff on coral and symbionts. These samples were collected at two time points: September 2017 (stress) and October 2017 (recover) to study the gene expression of Orbicella faveolata and Orbicella franksi. Also studied was Breviolum minutum, an algal symbiont which showed differentially expressed genes (DEGs) at different sampling time points. This may be the algal symbiont's response to the freshwater event (change in habitat). The purpose of our assignment was to study these DEGs at two time points, the stress and recovery. By utilizing arrayQualityMetrics, we were able to identify outliers. DESeq2 was leveraged to identify these DEGs and create a statistical model to study the expression patterns. 



**Conduct array quality metrics to detect and remove outliers**

```{r Packages, message=FALSE, warning=FALSE}
library(DESeq2) 
library(affycoretools)
library(arrayQualityMetrics)
library(genefilter)
library(Biobase)
library(ggplot2)
library(dplyr)
library(pheatmap)
library(vegan)
library(ggrepel)
library(tidyverse)
```

**read in iso2gene tab separated file**
```{r message=FALSE, warning=FALSE}
gg=read.table("Bmin_iso2gene.tab",sep="\t")
```

**read in counts** 
```{r message=FALSE, warning=FALSE}
countData <- read.table("allcounts_Harvey_Sym.txt") 
head(countData)
length(countData[,1])
```


**look for outliers**
```{r message=FALSE, warning=FALSE}
treat=c( "Recovery", "Recovery", "Recovery", "Stress", "Stress", "Stress") #creating a condition table
g=data.frame(treat)
g
colData= g
```

```{r echo=FALSE, fig.cap="**Figure X.** outliers"}
knitr::include_graphics('/Users/simminishad/Desktop/BI586/Assign2_proj4/arrayQualityMetrics report for e/out hm.png')
```
```{r echo=FALSE, fig.cap="**Figure X.** outliers"}
knitr::include_graphics('/Users/simminishad/Desktop/BI586/Assign2_proj4/arrayQualityMetrics report for e/out ma.png')
```

**Convert the counts data, treatment information, and desired statistical model into a DESeq object.**
```{r message=FALSE, warning=FALSE}
dds=DESeqDataSetFromMatrix(countData=countData,
                           colData = g,
                           design = ~treat)
vsd.ge=assay(vst(dds))
rl=vst(dds)
e=ExpressionSet(assay(rl), AnnotatedDataFrame(as.data.frame(colData(rl))))
arrayQualityMetrics(e, intgroup=c("treat"),force=T) 
```


**fav_stressC detected as outliers via the Distances between arrays method (Figure 2) and MA plot method (Figure 9). For the purpose of this assignment, we will include it, but in a real scenario we would exclude it.**


**one step DESeq that does the following: estimating size factors, estimating dispersions, gene-wise dispersion estimates, mean-dispersion relationship, final dispersion estimates, fitting model and testing.**
```{r message=FALSE, warning=FALSE}
dds<-DESeq(dds)
```

**These results allow us to make sure nothing abnormal is going on. The post-normalized counts are pvalues that correspond to see how differentially expressed a gene is between treatments. padj adjusts pvalues for multiple tests. The baseMean is average count for one comparison (whichever treatment group came first). log2FoldChange shows differentiation in counts between the basemean vs other treatment group. The recovery treatment group had avg of 5.57 counts; stress group had a 2fold change (logtransformed).**

```{r message=FALSE, warning=FALSE}
head(dds)
res<- results(dds) 
res
```



**This is the normalization method.**
```{r message=FALSE, warning=FALSE, fig.cap= "**Figure X.** The dispersion plot visualizes what DEseq2 is doing to our data; fitting it to this curve."}
plotDispEsts(dds, main="Wright et al. 2019 symbiont dispersion plot")
```


**Stress vs recovery pairwise comparisons**

**We can only do pairwise comparisons. Levels are used because gene expression is always relative; the first term is the treatment and the second term is the one that's relative (treatment 1st, control 2nd)**
```{r message=FALSE, warning=FALSE}
colData$Stress<-factor(colData$treat, levels=c("Stress","Recovery"))
```

```{r message=FALSE, warning=FALSE}
resStress <- results(dds, contrast=c("treat","Stress","Recovery"))
```

**False discovery rate pvalues are adjusted to < 10%**
```{r message=FALSE, warning=FALSE}
table(resStress$padj<0.01)
```

**Of the 5093 reads, p-adjusted reads are 6. p-adjusted values are used to look at numbers of differentially expressed genes - here we have 6 diferentially expressed genes.p-adjusted accounts for all the different times things are compared. Without adjustment we could get a false positive by chance.**

```{r message=FALSE, warning=FALSE}
summary(resStress)
```

**A lot of genes were removed due to low counts; 0.1% of genes upregulated in stress relative to recovery, and 0.43% were downregulated under stress relative to recovery**

**This is another way to look at the number of differentially expressed genes (6). The number of significantly differentially expressed genes excluding the no/low count genes.**

```{r message=FALSE, warning=FALSE}
nrow(resStress[resStress$padj<0.1 & !is.na(resStress$padj),])
```
  

```{r message=FALSE, warning=FALSE, fig.cap= "**Figure X.** MA plot shows the mean of normalized counts. The left is more lowly expressed genes abd right is higher expressed genes. The blue is all the differentially expressed genes. Negative fold change is downregulated genes and postitive fold change is upregulated genes."}
plotMA(resStress, main="Stress vs Recovery", ylim=c(-2,2))
```




**Our results are put in a data frame. 23 are upregulated and 98 are downregulated.**
```{r message=FALSE, warning=FALSE}
results <- as.data.frame(resStress)
head(results)
nrow(resStress[resStress$padj<0.1 & resStress$log2FoldChange > 0 & !is.na(resStress$padj),]) #no. genes upregulated
nrow(resStress[resStress$padj<0.1 & resStress$log2FoldChange < 0 & !is.na(resStress$padj),]) #no. genes downregulated
```


**A supplementary table of info.**
```{r message=FALSE, warning=FALSE}
write.table(resStress, file="StressvRecovery.txt", quote=F, sep="\t")
```

```{r message=FALSE, warning=FALSE}
cd <- read.table("StressvRecovery.txt")
head(cd)
```

###############################################################################################
##############################################################################

#*********what are pvals and rlogdata??

#--------------get pvals
#normal pval - get from running deseq and contrasting functions
#p val for one gene - if sig., it's diff expressed between treatment and control
valStress=cbind(resStress$pvalue, resStress$padj)
head(valStress)
colnames(valStress)=c("pval.Stress", "padj.Stress")
length(valStress[,1])
table(complete.cases(valStress))

valStress=cbind(resStress$pvalue, resStress$padj)
head(valStress)
colnames(valStress)=c("pval.Stress", "padj.Stress")
length(valStress[,1])
table(complete.cases(valStress))

######-------------make rlogdata and pvals table
#for heatmap, dont want colors corresponding to p value -- want it to correspond to actual counts
#for a PCoA, can't be based off P values, has to be differences in counts - need to normalize counts, can't just have raw counts (due to diffs in library size)
#transforms coutns so we can compare between samples
rlog=rlogTransformation(dds, blind=TRUE) 
rld=assay(rlog)
head(rld)
colnames(rld)=paste(colData$treat)
head(rld) #these numbers represent transformed counts - some are neg bc of the log transform
length(rld[,1])

rldpvals=cbind(rld,valStress, valStress)
head(rldpvals)
dim(rldpvals)
table(complete.cases(rldpvals)) #getting rid of p values that have NAs - data R wasn't able to calc a p value for (for many reasons: outlier, etc)

write.csv(rldpvals, "RLDandPVALS.csv", quote=F)
#putting counts and p values together



# 2 stress samples that act similarly, and one stress sample that has different gene expression...lots of genes doing same thing, or different samples respond differently
#if there's a strong signature, all stress samples would cluster together..here, 2/3 stress samples cluster, but third is odd
#the odd one out is the outlier?

#################################################################################
###########################heat map of sample distances

rldpvals <- read.csv(file="RLDandPVALS.csv", row.names=1)
head(rldpvals)
rld=rldpvals[,1:6]
head(rld)

colnames(rld)=paste(colData$treat)
head(rld)

library(RColorBrewer)

# Sample distance heatmap

library(gplots)
sampleDists <- dist(t(rld))
sampleDistMatrix <- as.matrix( sampleDists )
treat=c("Recovery", "Recovery", "Recovery", "Stress", "Stress", "Stress")
colnames(sampleDistMatrix)=paste(treat)
rownames(sampleDistMatrix)=paste(treat)

library("pheatmap")
heat.colors = colorRampPalette(rev(c("blue","yellow","red")),bias=0.3)(100)
pheatmap(sampleDistMatrix,color = heat.colors,cex=0.9,border_color=NA,cluster_rows=T,cluster_cols=T)
library(vegan)
library(ggplot2)
library(ggrepel)
library(tidyverse)
## Principal Components Analysis

Next, we run a Principal Components Analysis (PCA) to see how the gene expression of each of our 6 samples group together based on similarity.
```{r}
rld_t=t(rld)
pca <- prcomp(rld_t,center = TRUE)
li <- pca$sdev^2 / sum(pca$sdev^2)
pc1v <- round(li[1] * 100, 1)
pc2v <- round(li[2] * 100, 1)
pca_s <- as.data.frame(pca$x)
head(pca_s)
```

The above table shows us the coordinates of each sample on each of the 6 principal components. The principal components 1 through 6 explain decreasing amounts of variation in the data, with 1 explaining the most variation and 6 the least. We select out the first two principal components below so that we can plot the PCA on 2 axes. 

```{r}
pca_s <- pca_s[,c(1,2)]
pca_s$Samples = row.names(pca_s)
pca_s$treat=colData$treat
head(pca_s)
```
The above table contains our final coordinates for each sample on the principal component axes. We next plot the PCA below.

```{r, fig.cap="**Figure 2. Principal components analysis of gene expression of stress vs recovery samples.** PC1 explains 28.1% of the variance and PC2 explains 21.4% of the variance, for a total of 49.5% of variance explained by the first two principal components. The gene expression of the samples groups by stress vs recovery."}
cbPalette <- c("darkgoldenrod2",  "darkolivegreen3", "dodgerblue3")
ggplot(pca_s, aes(PC1, PC2, color = treat, pch = treat)) +
  geom_point(size=3) +
  #  geom_text_repel(aes(label=Samples)) +
  scale_colour_manual(values=cbPalette)+
  theme_bw() +
  # geom_density2d(alpha=.5)+
  geom_polygon(alpha=.2)+
  xlab(paste0("PC1: ",pc1v,"% variance")) +
  ylab(paste0("PC2: ",pc2v,"% variance")) 
```

From this plot, we can see that the symbiont gene expression of the samples groups by treatment (stress vs. recovery). Together, the first two principal components explain 49.5% of the variance in the data. The percentage of variance explained is likely low because we only included 6 samples in the analysis. 

To see if the clustering of gene expression of the two treatments on the principal component axes is significantly different, we run an analysis of variance with distance matrices. 
```{r, message = FALSE}
adonis(pca$x ~ treat, data = pca_s, method='eu', na.rm = TRUE)
```

This output tells us that the P-value ["Pr(>F)"] is 0.1. Thus, the two treatments are not significantly different, but are marginally different. However, because of our low sample size and strong visual clustering of treatments on the plotted PCA, we conclude that there is likely a difference in the gene expression of coral symbionts in the stressed versus recovery conditions. 

## Heatmaps

To create heatmaps to visualize the gene expression across our samples, we take the P-values and Rlog data, and read in the .csv with gene annotations. 

```{r, message=FALSE, warning = FALSE}
setwd("~/Desktop/BU/PhD/Spring_2021_classes/Ecological_genomics/Assign2/Project4")
rldpvals <- read.csv(file="RLDandPVALS.csv", row.names=1)
rld_site= rldpvals[,1:6]
gg=read.table("Bmin_iso2gene.tab",sep="\t", row.names=1)
nrow(rldpvals[rldpvals$padj.Stress<0.1& !is.na(rldpvals$padj.Stress),])
```

The above output displays how many differentially expressed genes we have using a p-value threshold of 0.1: 121 genes. We picked 0.1 as the P-value cutoff because we have very few samples, so we will not find many significantly differentially expressed genes with a very strict P-value. 

###Heatmap with all differentially expressed genes

121 differentially expressed genes is a manageable number of genes to visualize, so we created a heatmap with all the differentially expressed genes.

```{r, fig.cap="**Figure X. Heatmap of all differentially expressed genes between stress and recovery conditions.** Each row of rectangles represents one differentially expressed gene. The x axis shows the 6 samples analyzed. The legend displays the level of up- or downregulation. 2 (orange) indicates that the gene is more upregulated relative to the other treatment, while -2 (teal) indicates that the gene is more downregulated relative to the other treatment. 121 genes were differentially expressed with a p-value cutoff of 0.1. For the majority of differentially expressed genes, they were upregulated during recovery relative to stress conditions.}

p.val=0.1 #this is the False Discovery Rate cutoff
conds=rldpvals[rldpvals$padj.Stress<=p.val & !is.na(rldpvals$padj.Stress) & rldpvals$padj.Stress<=p.val & !is.na(rldpvals$padj.Stress),]
rld_data= conds[,c(1:6)]
library(pheatmap)
means=apply(rld_data,1,mean) # means of rows
explc=rld_data-means # subtracting them

ccol=colorRampPalette(rev(c("red","chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)

pheatmap(explc,cluster_cols=T,scale="row",color=col0, show_rownames = F)
```

This figure shows that for the majority of differentially expressed genes, they were upregulated during recovery relative to stress conditions. However, there is a small chunk of genes that were upregulated under stress relative to recovery conditions (in the bottom of the heatmap). 

###Heatmap with annotated genes

Next, we wanted to create a heatmap with annotated genes so we know which genes are being differentially expressed between stress and recovery conditions. Because a heatmap with 122 annotated genes would be overwhelming, we chose to create a heatmap with only the strongest differentially expressed genes. So, we changed our p-value cutoff to 0.01 to capture and annotate only the highly significant differentially expressed genes. 

First, we generated the heatmap with the p-value cutoff of 0.01 without annotations. 6 genes are differentially expressed with a p-value of 0.01.

```{r}
p.val=0.01 # changing the False Discovery Rate cutoff
conds=rldpvals[rldpvals$padj.Stress<=p.val & !is.na(rldpvals$padj.Stress) & rldpvals$padj.Stress<=p.val & !is.na(rldpvals$padj.Stress),]
rld_data= conds[,c(1:6)]
head(rld_data)
nrow(rld_data)
library(pheatmap)
means=apply(rld_data,1,mean) # means of rows
explc=rld_data-means # subtracting them

ccol=colorRampPalette(rev(c("red","chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)

pheatmap(explc,cluster_cols=T,scale="row",color=col0, show_rownames = F)

```

Next, we added gene annotations and set colors with the following code.

```{r, fig.cap="**Figure X. Heatmap of top 6 differentially expressed genes.** Each row of rectangles represents one differentially expressed gene. The x axis shows the 6 samples analyzed. The legend displays the level of up- or downregulation. 2 (orange) indicates that the gene is more upregulated relative to the other treatment, while -2 (teal) indicates that the gene is more downregulated relative to the other treatment. 6 genes were differentially expressed with a p-value cutoff of 0.01.}
ann = data.frame(cond = c('Recovery', 'Recovery', 'Recovery', 'Stress', 'Stress', 'Stress'))
rownames(ann) <- names(explc)

Var1        <- c("darkgoldenrod2","dodgerblue3")
names(Var1) <- c("Recovery", "Stress")
anno_colors <- list(cond = Var1)

pheatmap(as.matrix(explc),annotation_col=ann,annotation_colors=anno_colors,cex=1.2,color=col0,border_color=NA,clustering_distance_rows="correlation",clustering_distance_cols="correlation", show_rownames=T)
```
In this figure, we see that 3 of 6 genes are more upregulated in the recovery relative to the stress conditions. Two genes are highly upregulated in just one of the stressed samples. One gene is upregulated in all stressed samples relative to the recovery samples. 

